### ISSUE TITLE
[BUG] Tapir-zio catchAll handler is not triggered

### ISSUE THREAD

**Issue by KarelCemus** [softwaremill/tapir#3108]
Tapir version: 1.7.0

Scala version: 2.13

ZIO Http version: 3.0.0-RC2

**Describe the bug**

I am using `tapir` with ZIO interpreter and want to write `catchAll` handler to transform all uncaught exception into unified format (such as json). However, this handler is not being triggered for no type of failure. Neither defects nor uncaught runtime exceptions.

**How to reproduce?**

```
import sttp.tapir.PublicEndpoint
import sttp.tapir.ztapir._
import sttp.tapir.server.ziohttp.ZioHttpInterpreter
import zio._
import zio.http._

object Main extends ZIOAppDefault {

  def countCharacters(s: String): ZIO[Any, Nothing, Int] = {
    // ZIO.succeed(s.length) // this works correctly
    ZIO.dieMessage("A message on die")                                // this doesn't trigger catchAll
    // ZIO.succeed(throw new RuntimeException("Actual runtime failure")) // this doesn't trigger catchAll
  }

  val countCharactersEndpoint: PublicEndpoint[String, Unit, Int, Any] =
    sttp.tapir.endpoint.get.in(path[String]).out(plainBody[Int])

  val countCharactersHttp: HttpApp[Any, Response] =
    ZioHttpInterpreter()
      .toHttp(countCharactersEndpoint.zServerLogic(countCharacters))
      // it is expected that one of these two is triggered but none of these is
      .catchAllCauseZIO[Any, Response](ex => ZIO.succeed(Response.text("recovered in all cause")))
      .catchAllZIO(ex => ZIO.succeed(Response.text("recovered in all")))

  val run = Server.serve(countCharactersHttp).provide(Server.default)
}
```

It is expected to get one of `recovered in` messages but get "Internal Server Failure"` instead.

**Additional information**

When I write the same code for just `zio-http` without `tapir`, it works fine and `catchAll` is triggered, which is why I report it here.

**Comment by adamw** [comment]
Tapir includes an [exception interceptor](https://github.com/softwaremill/tapir/blob/8214653351a39c70b671944be65c07627c73ee94/server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionInterceptor.scala#L13) that is enable by default, and which transforms all errors/ exceptions into 500 responses. The best way to customise this is to customise the exception interceptor as [described here](https://tapir.softwaremill.com/en/latest/server/errors.html#customising-how-error-messages-are-rendered), either by specifying how all error messages should be rendered (including decode errors), or by providing a custom `ExceptionHandler` instance.

Alternatively, you can disable the interceptor altogether, but even then, I'm not sure if you'll see exceptions/defect caught by the code you wrote - it seems as if you are handling errors that might occur *during the construction of the routes*, not when *handling routes*.

**Comment by KarelCemus** [comment]
Thank you for your response, I'll check. If it turns out to solve the issue it would be great adding it into Tapir documentation.

> by the code you wrote - it seems as if you are handling errors that might occur during the construction of the routes, not when handling routes.

With ZIO only (without tapir), it catches errors occurred in handlers, I tested it before I opened this ticket.

**Comment by adamw** [comment]
> Thank you for your response, I'll check. If it turns out to solve the issue it would be great adding it into Tapir documentation.

Sure, I was hoping that the [error handling docs](https://tapir.softwaremill.com/en/latest/server/errors.html) cover this, but if you were expecting this information in another place, please let us know :)
 
> With ZIO only (without tapir), it catches errors occurred in handlers, I tested it before I opened this ticket.

Ah ok, it must be a feature of the API that I'm not aware of :)

**Comment by KarelCemus** [comment]
Thanks for your reply. I checked the documentation and the custom `ExceptionHandler` interceptor does the trick. [This documentation](https://tapir.softwaremill.com/en/latest/server/errors.html#customising-how-error-messages-are-rendered) shows an example for implementation based on Netty and there are no mentions of ZIO here or the exception handler in the ZIO part of the documentation but fortunately, there exists ZIO version of the same and I could do the following

```
ZioHttpInterpreter(ZioHttpServerOptions.customiseInterceptors.exceptionHandler(myExceptionHandler).options)
```

It is also worth mentioning that when I disabled the default exception handler through

```
ZioHttpInterpreter(ZioHttpServerOptions.customiseInterceptors[Any].exceptionHandler(None).options)
```

then the code I originally posted in the first post works as expected. Apparently, it is ok except that the default exception handler handles exceptions and thus the subsequent handler such as `catchAllZIO` has nothing to handle.

It might be worth adding this observation into the documentation but feel free to close this ticket as well.

Thanks for your help.
